pymongo/client_session.py

```diff
@@ -368,6 +377,97 @@ def _inherit_option(self, name, val):
             return val
         return getattr(self.client, name)
 
+    def with_transaction(self, callback, read_concern=None, write_concern=None,
+                         read_preference=None):
+        """Execute a callback in a transaction.
+
+        .. code-block:: python
+
+          def callback(session):
+              orders = session.client.db.orders
+              inventory = session.client.db.inventory
+              orders.insert_one({"sku": "abc123", "qty": 100}, session=session)
+              inventory.update_one({"sku": "abc123", "qty": {"$gte": 100}},
+                                   {"$inc": {"qty": -100}}, session=session)
+
+          with client.start_session() as session:
+              session.with_transaction(callback)
+
+        In an ideal scenario, `with_transaction` will start a transaction,
+        execute the callback once, and commit the transaction. In the event of
+        error, the commit or entire transaction may need to be retried and
+        thus the callback may be invoked multiple times. Even though the
+        callback may be invoked multiple times, `with_transaction` ensures
+        that the transaction will be committed at most once on the server.
+
+        In order to safeguard applications from infinite retry loops,
+        `with_transaction` will cease retrying the callback and
+        `commit_transaction` if it has exceeded a fixed timeout period of 120
+        seconds. This limit is non-configurable and is intentionally twice the
+        value of MongoDB's default for the  `transactionLifetimeLimitSeconds`
+        parameter (60 seconds). Applications that desire longer retry periods
+        may call `with_transaction` additional times as needed. Applications
+        that desire shorter retry periods should not use this method.
+
+        :Parameters:
+          - `callback`: The callback to run inside a transaction which must
+            accept a single argument, this session. Under certain error
+            conditions the callback may be run multiple times.
+          - `read_concern` (optional): The
+            :class:`~pymongo.read_concern.ReadConcern` to use for this
+            transaction.
+          - `write_concern` (optional): The
+            :class:`~pymongo.write_concern.WriteConcern` to use for this
+            transaction.
+          - `read_preference` (optional): The read preference to use for this
+            transaction. If ``None`` (the default) the :attr:`read_preference`
+            of this :class:`Database` is used. See
+            :mod:`~pymongo.read_preferences` for options.
```

Comment:
Should we use a single `opts` argument for `TransactionOptions` instead of passing `read_concern`, `write_concern`, and `read_preference` individually?

Personally I would prefer `def with_transaction(self, callback, opts=None)` because it avoids the need to add new keyword arguments every time we add a new transaction option but pymongo usually exposes individually options. Also, `start_transaction` accepts the individual options.

Reply:
Let's stick with the existing pattern.

Reply:
Sound good to me.